### PR TITLE
Created ObjectTest for testing Objects

### PR DESCRIPTION
**Pull request by Hedgemon4**

### CODE REVIEW

File: src/global/ObjectTest.java [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+                if (namesSet.contains(clause.getName())) {
+                    _fail("There is an issue with the test definition. Please contact an administrator.",
+                            "The name " + clause.getName() + " is already in use. Make sure all names are unique");
+                }
+                namesSet.add(clause.getName());
+            }
+            try {
+                clause.validateParams();
+            } catch (InvalidClauseException e) {
+                throw new InvalidClauseException("The parameter(s) for the " + methodName + " method clause at index " + i + " are invalid. " + e.getMessage());
```
Comment (keyvankhademi) [inline src/global/ObjectTest.java:476]: What will happen when we throw an error without `_fail`? What does the user see?

File: src/global/ObjectTest.java [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+            to check method console output using clauses. Note that random clauses and the placeHolder clause is not
+            currently supported for methods.
+         */
+
+        String output = methodOutput.toString();
+
+        if (methodTestSentence == null) {
+            assertEquals("", output,
+                    String.join(
+                            "",
```
Comment (keyvankhademi) [inline src/global/ObjectTest.java:423]: Why are we having an empty string at the beginning?

Comment (keyvankhademi) [reply src/global/ObjectTest.java:423]: oh that's what it joins it with.

File: src/global/ObjectTest.java [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+                                    "class does not have the correct modifiers."));
+                }
+            }
+            ByteArrayOutputStream methodOutput = new ByteArrayOutputStream();
+            System.setOut(new PrintStream(methodOutput));
+            Object output = objectMethodInvoke.invoke(object, args);
+            methodOutputFollowsCorrectStructure(methodName, methodTestSentence, incorrectMethodStructureErrorMessage, methodOutput);
+            return output;
+        } catch (NoSuchMethodException e) {
+            fail(String.join("", "Your ", objectClass.getSimpleName(),
```
Comment (keyvankhademi) [inline src/global/ObjectTest.java:377]: Why do we have both `fail` and `_fail`? Whats the difference?

**Review by keyvankhademi — DISMISSED** [review]
This is great. Seems to be error-proof too.

File: src/global/ObjectTest.java (hunk 3 above)
Comment (Hedgemon4) [reply src/global/ObjectTest.java:377]: _fail allows us to use two messages. The first is the error message the user will see, and the second logs a developer message to the console which tells them if they made a mistake with something, or gives them more description of the error.

File: src/global/ObjectTest.java (hunk 1 above)
Comment (Hedgemon4) [reply src/global/ObjectTest.java:476]: I think they will see this, but this should be changed to fail.

Comment (Hedgemon4) [reply src/global/ObjectTest.java:476]: This is actually a developer error if the clause is not defined correctly, so the student should never see this error. However, I will change it to use _fail anyways.

Comment (Hedgemon4) [reply src/global/ObjectTest.java:476]: Unless there is an error with our test, but they should be tested as they are put on the website to make sure this doesn't happen.

File: src/global/ObjectTest.java [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+    }
+
+    public Object createArray(int length) {
+        if (length < 0)
+            _fail("Error with test definition. Please contact a test administrator.",
+                    "Must pass a positive length to initialize an array.");
+        return Array.newInstance(this.objectClass, length);
+    }
+
+    public Object createArray(int length, Object[][][] arguments) throws Throwable {
```
Comment (keyvankhademi) [inline src/global/ObjectTest.java:497]: Let's add a method that creates identical elements with the same input as well.

**Review by keyvankhademi — APPROVED** [review]
Well done.

**Pull request merged**